controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CommitTransactionTask.java

```diff
@@ -0,0 +1,174 @@
+        log.debug("Attempting to commit available transactions on epoch {} on stream {}/{}", event.getEpoch(), event.getScope(), event.getStream());
+
+        return streamMetadataStore.getActiveEpoch(scope, stream, context, false, executor).thenComposeAsync(pair -> {
+            if (epoch < pair.getKey()) {
+                log.debug("Epoch {} on stream {}/{} is already complete.", epoch, scope, stream);
+                return CompletableFuture.completedFuture(null);
+            } else if (epoch == pair.getKey()) {
+                // If the transaction's epoch is same as the stream's current epoch, commit it.
+                return tryCommitTransactions(scope, stream, epoch, context, this.streamMetadataTasks.retrieveDelegationToken());
+            } else {
+                // If commit request if from higher epoch, throw OperationNotAllowed. This will result in event being posted back.
+                String message = String.format("Transaction on old epoch should complete before allowing commits on new epoch %d on stream %s/%s",
+                        epoch, scope, stream);
+                throw StoreException.create(StoreException.Type.OPERATION_NOT_ALLOWED, message);
+            }
+        }).whenCompleteAsync((result, error) -> {
+            if (error != null) {
+                log.error("Exception while attempting to committ transaction on epoch {} on stream {}/{}", epoch, scope, stream);
+            } else {
+                log.debug("Successfully committed transactions on epoch {} on stream {}/{}", epoch, scope, stream);
+                if (processedEvents != null) {
+                    processedEvents.offer(event);
+                }
+            }
+        }, executor);
+    }
+
+    @Override
+    public CompletableFuture<Void> writeBack(CommitEvent event) {
+        return streamMetadataTasks.writeEvent(event);
+    }
+
+    private CompletableFuture<Void> tryCommitTransactions(final String scope,
+                                                  final String stream,
+                                                  final int epoch,
+                                                  final OperationContext context,
+                                                  final String delegationToken) {
+        // if node already exists and doesnt match event.transactionId, throw operation not allowed. dont worry,
+        // it will be posted back in the stream and retried later. Generally if a transaction commit starts, it will come to
+        // an end.. but during failover,
+        // once we have created the node, we are guaranteed that it will be only that transaction that will be getting
+        // committed at that time.. so post failover even when we recover, we will handle one transaction at a time.
+        CompletableFuture<List<UUID>> txnListFuture = streamMetadataStore.getTxnCommitList(scope, stream, context, executor)
+                .thenCompose(record -> {
```

Comment:
`thenComposeAsync`, with executor passed in

Reply:
replaced thenComposeAsync with thenCompose.
ThenComposeAsync came because of copy pasting that part of the code from the original class (CommitEventProcessor.java) which has been replaced by this new class.